### DISCUSSION TITLE
Typology of results from orienteering software

### DISCUSSION THREAD

**Discussion by jarufe** [oreplay/oreplay#4] [General]
In order to study the kind of data provided for the most common orienteering software. The link below allows users to edit the document. The goal is to share and to complete the information between all of the collaborators.
[Link to Google Docs](https://docs.google.com/document/d/1uhA4RZE_1-iVBxBgNSrXkhQd9R0jGnqYu_nK8RYKigQ/edit?usp=sharing)

**Comment by freefri** [comment]
We need to define how to calculate multi stage event scores and multi event rankings. Maybe we can use some predefined constants to define if results should be computed as multi stage or ranking and which criteria should apply. Eligibility fields could be used for this purpose.

Calculation of position is hard and could result in discrepancies between timing software and oreplay. I would suggest to keep fields like time behind ~~or position~~, calculated dynamically by oreplay. This would allow that adding one new finisher in 1st place on a class does not need a change of all stored time behind  and positions. This could be problematic thoug, if we want to display results where the system does not query and runners in one class. For example if we want to display all possitions and behind times for all runner in one club. This could be solved but storing those fields in some kind of persistent cache.

Regarding XML output: great job summarizing. I think eligibility is a key field to be used. Sad that done software does no stick to standard iof v3. We would need to try to convince software makers to adapt some changes, or directly ask them to implement API connection (to avoid implementing our own desktop client). We could define our own standard based on iof v3, but maybe simplified and with some small additions. Otherwise we would need to work with CSV files.

**Comment by jarufe** [comment]
Timestamp ISO 8601 format for start times or finish times. OE, OS, Sitiming don't use the UTC zone in the formatting of their times (Z, +01:00, -02:00, etc). One example is 2014-07-06T13:09:00.000, instead of 2014-07-06T13:09:00.000+02:00

In our model, the event table supports the definition of the UTC Zone, so we could display times (mainly the start times) adapted to the UTC Zone of an spectator. For example, an spectator in Spain (UTC Zone +02:00) viewing an event in Portugal (UTC Zone +01:00). For a runner starting his race at Portugal's time 10:30:00, an spectator in Spain would start following that runner's race at Spain's time 11:30:00.

If I'm not wrong, Liveresults don't take this in account and times are shown in event's local time.

For viewing, if I don't show the UTC part of the ISO 8601 date, apparently it would be Ok. But, attention, if an organiser in Portugal creates an event in OReplay, as OReplay server is located in Spain, when parsing XML and CSV data with times, by default they are converted to Date type values using the UTC Zone of the server, so the internal value of the date is not going to be the same as its representation.
Probably, it should be good that the creator of an event declares its UTC Zone. Then, when importing times, conversion would be done to save a Date type that is the same in our server than locally at the event. And when viewing data, another conversion would be done to represent that Date in the spectator's UTC Zone.

And, Attention to another thing, the difference of the UTC Zone values when changing between Winter and Summer times. For example, Spain is said to be at +02:00, but is at +01:00 in winter.

Or, don't do anything and keep things as simple as their are in Liveresultat.

**Reply by freefri** [reply]
I think the best option would be that the desktop client will always set a time zone based on the os timezone. The desktop client should auto convert the time zone and send altura the time with timezone.

Another option would be: When creating the event, we should take the local time zone of the users browser by default, and store it as part of event entity. 
When uploading the results we would assume no time zone will be converted to the event time zone every time. But I think this is a worst approach

**Reply by freefri** [reply]
@jarufe Just to clarify after our last meeting with @sergiogpajares and @udernaeb.

I think the best is that all dates are always stored in UTC in the server. The desktop client should send the times in local time always and the conversion will be done in the server.

Times will be sent from the API to the frontend in the ISO standard always with timezone added like +xx.xx so the web app can automatically translate it to the timezone of the end user. We should use some nice library or some browser api for this, read: https://momentjs.com/docs/#/-project-status/ (maybe @jessicapfernandez has some ideas already)
